pkg/clusteragent/admission/mutate/autoinstrumentation/injector.go

```diff
@@ -0,0 +1,160 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver
+
+package autoinstrumentation
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+var sourceVolume = volume{
+	Volume: corev1.Volume{
+		Name: volumeName,
+		VolumeSource: corev1.VolumeSource{
+			EmptyDir: &corev1.EmptyDirVolumeSource{},
+		},
+	},
+}
+
+var v1VolumeMount = sourceVolume.mount(corev1.VolumeMount{
+	MountPath: mountPath,
+})
+
+var v2VolumeMountInjector = sourceVolume.mount(corev1.VolumeMount{
+	MountPath: "/opt/datadog-packages/datadog-apm-inject",
+	SubPath:   "opt/datadog-packages/datadog-apm-inject",
+})
+
+var v2VolumeMountLibrary = sourceVolume.mount(corev1.VolumeMount{
+	MountPath: "/opt/datadog/apm/library",
+	SubPath:   "opt/datadog/apm/library",
+})
+
+type injector struct {
+	image      string
+	registry   string
+	injected   bool
+	injectTime time.Time
+}
+
+func (i *injector) initContainer() initContainer {
+	var (
+		name  = "datadog-init-apm-inject"
+		mount = corev1.VolumeMount{
+			MountPath: "/datadog-inject",
+			SubPath:   v2VolumeMountInjector.SubPath,
+			Name:      v2VolumeMountInjector.Name,
+		}
+		tsFilePath = mount.MountPath + "/c-init-time." + name
+	)
+	return initContainer{
+		Prepend: true,
+		Container: corev1.Container{
+			Name:    name,
+			Image:   i.image,
+			Command: []string{"/bin/sh", "-c", "--"},
+			Args: []string{
+				fmt.Sprintf(
+					`cp -r /%s/* %s && echo $(date +%%s) >> %s`,
+					mount.SubPath, mount.MountPath, tsFilePath,
+				),
+			},
+			VolumeMounts: []corev1.VolumeMount{mount},
+		},
+	}
+}
+
+func (i *injector) requirements() libRequirement {
+	return libRequirement{
+		initContainers: []initContainer{i.initContainer()},
+		volumes:        []volume{sourceVolume},
+		volumeMounts:   []volumeMount{v2VolumeMountInjector.readOnly().prepended()},
```

Comment:
we probably also should be mounting in `/opt/datadog-packages/datadog-apm-inject/stable/inject/ld.so.preload` to `/etc/ld.so.preload` so that we instrument things that don't have the LD_PRELOAD environment variable propagated. This could happen if there's a bootstrapping shell script that starts the application with a different environment.

Reply:
@Kyle-Verhoog I'm going to branch off of this one and test that separately. Do we need the env var if we're the path?

Reply:
https://github.com/DataDog/datadog-agent/pull/28321 going to test here

Reply:
I think we'd end up keeping both the env var and the path?

Reply:
Another thing here is that if there's a shell script that strips the LD_PRELOAD env var, should we respect that? It would work on alpine and not work on glibc and that'd be odd. But this is an inconsistency with what we're doing on host 🤷